ultratrace2/model/files/loaders/base.py

```diff
@@ -0,0 +1,58 @@
+import logging
+
+from abc import ABC, abstractmethod
+from PIL import Image  # type: ignore
+
+
+logger = logging.getLogger(__name__)
+
+
+class FileLoadError(Exception):
+    pass
+
+
+class FileLoaderBase(ABC):
+    def __new__(cls, path: str):
+        try:
+            return cls.from_file(path)
+        except FileLoadError as e:
+            logger.error(e)
+            return None
+
+    def __init__(self, path: str):
+        self.path = path
+
+    def __repr__(self):
+        return f"{type(self).__name__}({self.path})"
+
+    @classmethod
+    @abstractmethod
+    def from_file(cls, path: str) -> "FileLoaderBase":
```

Comment:
This won't typecheck properly for subclasses which return an instance of the subclass. You need to define a `TypeVar` above:
```python
Self = TypeVar("Self", bound="FileLoaderBase")
```
then constrain the return type using `cls`:
```suggestion
    def from_file(cls: Type[Self], path: str) -> Self:
```
see: https://github.com/python/mypy/issues/1212#issuecomment-219250132

Reply:
> This won't typecheck properly for subclasses which return an instance of the subclass. You need to define a `TypeVar` above:
> 
> ```python
> Self = TypeVar("Self", bound="FileLoaderBase")
> ```
> 
> then constrain the return type using `cls`:

Ah, this is what I was trying to puzzle through yesterday.  Thanks.

> see: [python/mypy#1212 (comment)](https://github.com/python/mypy/issues/1212#issuecomment-219250132)

lmao, straight from GvR, nice.

Reply:
Fixed in a452e98